### PR TITLE
deps: bump typescript to 4.1.3 and linting deps

### PR DESCRIPTION
**Pull request by fedeci**
## Description
This is useful to take advantage of all new typescript features like abstract classes.

<s>__`v4.1.3` is still not supported by `@typescript-eslint/typescript-estree`__</s>
## To-do list

- [x] I have added tests for what I changed.

- [x] This pull request is ready to merge.

### CODE REVIEW

**Review by gustavohenke — APPROVED** [review]
yay thanks!

**Comment by gustavohenke** [comment]
do you reckon we can release a v7.0.0 now that TS is compiled to work with v4, or is there any other breaking change we can fit in it?

**Comment by fedeci** [comment]
If I am not wrong there are other PRs waiting to be merged in a major release, but we could include some other breaking changes to this (e.g. wildcards and oneOf rework). I don't even know if this needs a major version as it is not expected to break anything.

File: src/chain/context-runner-impl.spec.ts [hunk 1]
```diff
@@ -78,13 +78,13 @@ it('runs items on the stack with required data', async () => {
 });
 
 it('runs items on the stack in order', async () => {
-  let item1Resolve = () => {};
+  let item1Resolve: Function = () => {};
```
Comment (gustavohenke) [inline src/chain/context-runner-impl.spec.ts:81]: Just for curiosity -- why?

**Review by gustavohenke — APPROVED** [review]

File: src/chain/context-runner-impl.spec.ts (hunk 1 above)
Comment (fedeci) [reply src/chain/context-runner-impl.spec.ts:81]: IDK why but TS was complaining

**Pull request merged**